src/mlpack/core/data/load_image_impl.hpp

```diff
@@ -14,154 +16,158 @@
-
-  for (size_t i = 1; i < files.size() ; ++i)
-  {
-    arma::Mat<unsigned char> colImg(tmpMatrix.colptr(i), tmpMatrix.n_rows, 1,
-        false, true);
-    status = LoadImage(files[i], colImg, info, fatal);
-
-    if (!status)
-      return false;
-  }
-
-  matrix = arma::conv_to<arma::Mat<eT>>::from(tmpMatrix);
-  return true;
-}
-
-inline bool LoadImage(const std::string& filename,
-                      arma::Mat<unsigned char>& matrix,
-                      ImageInfo& info,
-                      const bool fatal)
-{
-  unsigned char* image;
-
-  if (!ImageFormatSupported(filename))
+  if (!opts.ImageFormatSupported(files.back()))
   {
-    std::ostringstream oss;
-    oss << "Load(): file type " << Extension(filename) << " not supported. ";
-    oss << "Currently it supports:";
-    auto x = LoadFileTypes();
+    std::stringstream oss;
+    oss << "Load(): file type " << Extension(files.back())
+        << " isn't supported. Currently it supports: ";
+    auto x = opts.LoadFileTypes();
     for (auto extension : x)
       oss << " " << extension;
-    oss << "." << std::endl;
-
-    if (fatal)
-    {
-      Log::Fatal << oss.str();
-    }
-    else
-    {
-      Log::Warn << oss.str();
-    }
-
-    return false;
+    mlpackException(oss, opts);
   }
 
   // Temporary variables needed as stb_image.h supports int parameters.
   int tempWidth, tempHeight, tempChannels;
-
-  // For grayscale images.
-  if (info.Channels() == 1)
+  // I think we can get rid of these two if we will be able to assign directly
+  // to an armadillo matrix, but I doubt this is possible
+  unsigned char* imageBuf;
+  float* floatImageBuf;
+  // Check dimension if provided or not
+  // if not, then call the function on the first image to get the dimension
+  // Note that, similar to the Resize function, we need to document that the
+  // user cannot load images with different dimension in the same matrix.
+  if (opts.Width() == 0 || opts.Height() == 0)
   {
-    image = stbi_load(filename.c_str(), &tempWidth, &tempHeight, &tempChannels,
-        STBI_grey);
+    imageBuf = stbi_load(files.front().c_str(), &tempWidth, &tempHeight,
+        &tempChannels, STBI_rgb);
+    if (!imageBuf)
+    {
+      std::stringstream oss;
+      oss << "Load(): failed to load image '" << files.front() << "': "
+              << stbi_failure_reason();
+    
+      mlpackException(oss, opts);
+    }
+    
+    opts.Width() = tempWidth;
+    opts.Height() = tempHeight;
+    opts.Channels() = tempChannels;
   }
+
+  dimension = opts.Width() * opts.Height() * opts.Channels();
+
+  arma::Mat<unsigned char> images;
+  arma::Mat<float> floatImages;
+
+  if constexpr (std::is_same<eT, float>::value)
+    floatImages.set_size(dimension, files.size());
   else
-  {
-    image = stbi_load(filename.c_str(), &tempWidth, &tempHeight, &tempChannels,
-        STBI_rgb);
-  }
+    images.set_size(dimension, files.size());
 
-  if (!image)
+  for (size_t i = 0; i < files.size(); ++i)
```

Comment:
This will always reload the first file.  I think it is better to either split the first iteration out, or check internally in the loop whether it is the first iteration and set the parameters of `opts` accordingly.